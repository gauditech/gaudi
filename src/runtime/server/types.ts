import { NextFunction, Request, Response } from "express";

export type ServerHttpMethods = "all" | "get" | "post" | "put" | "patch" | "delete";

export type ServerRequestHandler =
  | ((request: Request, response: Response) => void | Promise<void>)
  | ((req: Request, resp: Response, next: NextFunction) => void | Promise<void>);

export type EndpointConfig = {
  path: string;
  method: ServerHttpMethods;
  handlers: ServerRequestHandler[];
};

export type ServerMiddlewareNextFn = (err?: unknown) => void;
