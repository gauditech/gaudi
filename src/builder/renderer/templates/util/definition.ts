import { Definition, PathFragment } from "@src/types/definition";

/*
 * Template helper functions for Gaudi definition
 */

export function modelDbName(modelRefKey: string, definition: Definition) {
  return (definition.models || []).find((model) => model.refKey === modelRefKey)?.dbname;
}

export function fieldDbName(modelRefKey: string, fieldRefKey: string, definition: Definition) {
  return (definition.models || [])
    .find((model) => model.refKey === modelRefKey)
    ?.fields.find((field) => field.refKey === fieldRefKey)?.dbname;
}

export function typeToDbType(type: string) {
  if (type === "serial") {
    return "Int";
  } else if (type === "integer") {
    return "Int";
  } else if (type === "text") {
    return "String";
  } else if (type === "boolean") {
    return "Boolean";
  }

  throw "UNHANDLED_GAUDI_TYPE";
}

export function buildEndpointPath(parts: PathFragment[]): string {
  return parts.reduce((accum, part) => {
    const pathPart =
      part.type === "numeric" || part.type === "text" ? `:${part.varname}` : part.value;
    return `${accum}/${pathPart}`;
  }, "");
}
