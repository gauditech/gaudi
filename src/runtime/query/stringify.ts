import { source } from "common-tags";
import _ from "lodash";

import { NamePath, selectToSelectable } from "./build";

import { getRef, getTargetModel } from "@src/common/refs";
import { BinaryOperator } from "@src/types/ast";
import {
  Definition,
  FilterDef,
  ModelDef,
  QueryDef,
  QueryDefPath,
  SelectFieldItem,
  SelectableItem,
} from "@src/types/definition";

// FIXME this should accept Queryable
export function queryToString(def: Definition, q: QueryDef): string {
  switch (q.from.kind) {
    case "model": {
      const { value: model } = getRef<"model">(def, q.from.refKey);
      return source`
      SELECT ${selectableToString(def, selectToSelectable(q.select))}
      FROM "${model.dbname}" AS ${namePathToAlias([model.name])}
      ${q.joinPaths.map((j) => joinToString(def, j))}
      WHERE ${filterToString(q.filter)}`;
    }
    case "query":
      return source`
      SELECT ${selectableToString(def, selectToSelectable(q.select))}
      FROM (${queryToString(def, q.from.query)}) AS ${namePathToAlias(q.from.query.fromPath)}
      ${q.joinPaths.map((j) => joinToString(def, j))}
      WHERE ${filterToString(q.filter)}`;
  }
}

function selectableToString(def: Definition, select: SelectableItem[]) {
  return select
    .map((item) => {
      const { value: field } = getRef<"field">(def, item.refKey);
      return `${namePathToAlias(_.initial(item.namePath))}.${field.dbname} AS ${item.alias}`;
    })
    .join(", ");
}

function namePathToAlias(namePath: NamePath): string {
  return `"${namePath.join(".")}"`;
}

function filterToString(filter: FilterDef): string {
  if (filter === undefined) return "TRUE = TRUE";
  switch (filter.kind) {
    case "literal": {
      switch (filter.type) {
        case "boolean":
          return filter.value ? "TRUE" : "FALSE";
        case "null":
          return "NULL";
        case "text":
          return `'${filter.value}'`;
        case "integer":
          return filter.value.toString();
      }
    }
    // Due to bug in eslint/prettier, linter complains that `break` is expected in the case "literal"
    // Since inner switch is exaustive, break is unreachable so prettier deletes it
    // eslint-disable-next-line no-fallthrough
    case "alias": {
      const np = filter.namePath.slice(0, filter.namePath.length - 1);
      const f = filter.namePath.at(filter.namePath.length - 1);
      return `${namePathToAlias(np)}.${f}`;
    }
    case "binary": {
      const fstr = `${filterToString(filter.lhs)} ${opToString(filter.operator)} ${filterToString(
        filter.rhs
      )}`;
      if (filter.operator === "or") {
        return `(${fstr})`;
      }
      return fstr;
    }
    case "variable": {
      return `:${filter.name}`;
    }
  }
}

function opToString(op: BinaryOperator): string {
  switch (op) {
    case "is":
      return "=";
    case "is not":
      return "<>";
    default:
      return op.toUpperCase();
  }
}

function getQuerySource(def: Definition, q: QueryDef): ModelDef {
  switch (q.from.kind) {
    case "model":
      return getRef<"model">(def, q.from.refKey).value;
    case "query": {
      return getQuerySource(def, q.from.query);
    }
  }
}

function joinToString(def: Definition, join: QueryDefPath): string {
  const model = getTargetModel(def.models, join.refKey);
  const joinNames = getJoinNames(def, join.refKey);
  const joinMode = join.joinType === "inner" ? "JOIN" : "LEFT JOIN";
  const joinFilter: FilterDef = {
    kind: "binary",
    operator: "is",
    lhs: { kind: "alias", namePath: [..._.initial(join.namePath), joinNames.from] },
    rhs: { kind: "alias", namePath: [...join.namePath, joinNames.to] },
  };

  let src: string;
  if (join.kind === "query") {
    const { value: query } = getRef<"query">(def, join.refKey);
    const sourceModel = getQuerySource(def, query);
    // extend select
    const conn = mkJoinConnection(sourceModel);

    const retModel = getRef<"model">(def, query.retType).value;
    const fields = retModel.fields.map(
      (f): SelectableItem => ({
        kind: "field",
        refKey: f.refKey,
        alias: f.name,
        name: f.name,
        namePath: [...query.fromPath, f.name],
      })
    );
    src = source`(
      ${queryToString(def, { ...query, select: [...query.select, ...fields, conn] })})`; // FIXME should remove query.select?
  } else {
    src = `"${model.dbname}"`;
  }

  return source`
  ${joinMode}
  ${src} AS ${namePathToAlias(join.namePath)}
  ON ${filterToString(joinFilter)}
  ${join.joinPaths.map((j) => joinToString(def, j))}`;
}

export function mkJoinConnection(model: ModelDef): SelectFieldItem {
  return {
    kind: "field",
    refKey: `${model.name}.id`,
    alias: '"__join_connection"',
    name: "id",
    namePath: [model.name, "id"],
  };
}

function getJoinNames(def: Definition, refKey: string): { from: string; to: string } {
  const prop = getRef(def, refKey);
  switch (prop.kind) {
    case "reference": {
      const reference = prop.value;
      const { value: field } = getRef<"field">(def, reference.fieldRefKey);
      const { value: refField } = getRef<"field">(def, reference.toModelFieldRefKey);

      return { from: field.name, to: refField.name };
    }
    case "relation": {
      const relation = prop.value;
      const { value: reference } = getRef<"reference">(def, relation.throughRefKey);
      const { value: field } = getRef<"field">(def, reference.fieldRefKey);
      const { value: refField } = getRef<"field">(def, reference.toModelFieldRefKey);

      return { from: refField.name, to: field.name };
    }
    case "query": {
      return { from: "id", to: '"__join_connection"' };
    }
    default:
      throw new Error(`Kind ${prop.kind} not supported`);
  }
}
