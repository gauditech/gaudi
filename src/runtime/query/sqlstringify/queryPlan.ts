import _ from "lodash";

import { getRef } from "@src/common/refs";
import { UnreachableError, assertUnreachable, ensureEqual, ensureNot } from "@src/common/utils";
import { VarContext, getTypedPath } from "@src/composer/utils";
import { Definition, FunctionName, QueryDef, TypedExprDef } from "@src/types/definition";

export type ShallowDep = {
  path: string[];
  aggregate: "sum" | "count" | null;
};
export type ShallowDeps = ShallowDep[];

/**
 * I don't know what this is yet!
 */

type NamePath = string[];

export type QueryPlan = {
  entry: string;
  // namePath: NamePath;
  joins: QueryPlanJoin[];
  groupBy: NamePath[];
  filter?: QueryPlanExpression;
  select?: Record<string, QueryPlanExpression>; // key is 'alias'
  orderBy?: [NamePath, "asc" | "desc"][];
  limit?: number;
  offset?: number;
};

export type JoinWithSubquery = {
  kind: "subquery";
  joinType: "inner" | "left";
  plan: QueryPlan;
  joins: QueryPlanJoin[];
};

export type InlineJoin = {
  kind: "inline";
  joinType: "inner" | "left";
  target: string;
  joins: QueryPlanJoin[];
};

export type QueryPlanJoin = JoinWithSubquery | InlineJoin;

export type QueryPlanExpression =
  | {
      kind: "literal";
      type: "boolean" | "null" | "text" | "integer";
      value: unknown;
    }
  | {
      kind: "alias";
      value: NamePath;
    }
  | { kind: "function"; fnName: FunctionName; args: QueryPlanExpression[] }
  | { kind: "variable"; name: string };

export function collectShalowDeps(q: QueryDef): ShallowDeps {
  // collect select
  const selectDeps = q.select.map((s): ShallowDep => {
    switch (s.kind) {
      case "computed":
      case "field": {
        return { path: s.namePath, aggregate: null };
      }
      default: {
        throw new UnreachableError(`"${s.kind}" is not a valid selectable item!`);
      }
    }
  });

  const filterDeps = expressionToShallowDeps(q.filter);

  const allDeps = [...selectDeps, ...filterDeps];
  return _.uniqBy(allDeps, (d) => `${d.aggregate}|${d.path.join(".")}`);
}

function expressionToShallowDeps(exp: TypedExprDef): ShallowDeps {
  switch (exp?.kind) {
    case undefined:
    case "variable":
    case "literal": {
      return [];
    }
    case "alias": {
      return [{ path: exp.namePath, aggregate: null }];
    }
    case "function": {
      switch (exp.name) {
        case "count":
        case "sum": {
          const target = expressionToShallowDeps(exp.args[0]);
          ensureEqual(target.length, 1);
          return [{ path: target[0].path, aggregate: exp.name }];
        }
        default: {
          // it's an actual function
          return exp.args.flatMap(_.unary(expressionToShallowDeps));
        }
      }
    }
    default: {
      return assertUnreachable(exp);
    }
  }
}

function expressionToExpandedDeps(
  def: Definition,
  exp: TypedExprDef,
  ctx: VarContext,
  shallow = false
): ShallowDeps {
  switch (exp?.kind) {
    case undefined:
    case "variable":
    case "literal": {
      return [];
    }
    case "alias": {
      const tpath = getTypedPath(def, exp.namePath, ctx);
      ensureNot(tpath.leaf, null);
      const ref = getRef(def, tpath.leaf?.refKey);
      switch (ref.kind) {
        case "field": {
          return [{ path: exp.namePath, aggregate: null }];
        }
        case "computed": {
          if (shallow) {
            return [{ path: exp.namePath, aggregate: null }];
          }
          return expressionToExpandedDeps(def, ref.exp, ctx);
        }
        default: {
          throw new UnreachableError(`${ref.kind} is not allowed query atom`);
        }
      }
    }
    case "function": {
      switch (exp.name) {
        case "count":
        case "sum": {
          const target = expressionToShallowDeps(exp.args[0]);
          ensureEqual(target.length, 1);
          return [{ path: target[0].path, aggregate: exp.name }];
        }
        default: {
          return exp.args.flatMap(_.unary(expressionToShallowDeps));
        }
      }
    }
  }
}

// export function buildQueryPlan(q: QueryDef): QueryPlan {
//   const atoms = collectShalowDeps(q);
//   const expandedAtoms = collectExpandedDeps(q);
// }
