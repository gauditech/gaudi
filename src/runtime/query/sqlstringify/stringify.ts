import _ from "lodash";

import { NamePath } from "../build";

import { QueryPlan, QueryPlanExpression, QueryPlanJoin } from "./queryPlan";

import { assertUnreachable } from "@src/common/utils";

function namepathToQuotedPair(npath: NamePath): string {
  return `"${_.initial(npath).join(".")}"."${_.last(npath)}"`;
}

function namepathToQuoted(npath: NamePath): string {
  return `"${npath.join(".")}"`;
}

function exprToString(expr: QueryPlanExpression): string {
  function stringifyOp(lhs: QueryPlanExpression, rhs: QueryPlanExpression, op: string): string {
    return `${exprToString(lhs)} ${op.toUpperCase()} ${exprToString(rhs)}`;
  }
  function stringifyFn(name: string, args: QueryPlanExpression[]): string {
    return `${name}(${args.map(_.unary(exprToString)).join(", ")})`;
  }
  switch (expr.kind) {
    case "alias": {
      return namepathToQuotedPair(expr.value);
    }
    case "literal": {
      return expr.value as string; // FIXME
    }
    case "variable": {
      return `:${expr.name}`;
    }
    case "function": {
      switch (expr.fnName) {
        case "<":
        case ">":
        case ">=":
        case "<=":
        case "/":
        case "*":
        case "in":
        case "not in":
        case "and": {
          return stringifyOp(expr.args[0], expr.args[1], expr.fnName);
        }
        case "+":
        case "-":
        case "or": {
          return `(${stringifyOp(expr.args[0], expr.args[1], expr.fnName)})`;
        }
        case "is": {
          return stringifyOp(expr.args[0], expr.args[1], "=");
        }
        case "is not": {
          return stringifyOp(expr.args[0], expr.args[1], "<>");
        }
        case "length": {
          return stringifyFn("char_length", expr.args);
        }
        case "count":
        case "sum":
        case "lower":
        case "upper":
        case "now":
        case "concat": {
          return stringifyFn(expr.fnName, expr.args);
        }
        case "stringify":
        case "cryptoCompare":
        case "cryptoHash":
        case "cryptoToken":
          throw new Error(`Expression "${expr.fnName}" cannot be used in queries.`);
        default:
          return assertUnreachable(expr.fnName);
      }
    }
  }
}

function joinsToString(joins: QueryPlanJoin[]): string {
  return joins
    .map((join) => {
      switch (join.kind) {
        case "subquery": {
          const subquery = queryPlanToString(join.plan);
          return `
    JOIN (${subquery})
    AS ${namepathToQuoted(join.namePath)}
    ON ${namepathToQuotedPair(join.joinOn[0])} = ${namepathToQuotedPair(join.joinOn[1])}
    ${joinsToString(join.plan.joins)}`;
        }
        case "inline": {
          return `
    JOIN ${join.modelName.toLowerCase()} AS ${namepathToQuoted(join.namePath)}
    ON ${namepathToQuotedPair(join.joinOn[0])} = ${namepathToQuotedPair(join.joinOn[1])}
    ${joinsToString(join.joins)}
    `;
        }
      }
    })
    .join("\n");
}

export function queryPlanToString(plan: QueryPlan): string {
  const selectFrag = plan.select
    ? Object.entries(plan.select)
        .map(([alias, expr]) => `${exprToString(expr)} AS "${alias}"`)
        .join(", ")
    : "*";
  const limitFrag = plan.limit ? `LIMIT ${plan.limit}` : "";
  const orderFrag = plan.orderBy
    ? `ORDER BY ${plan.orderBy.map(
        ([path, dir]): string => `${namepathToQuotedPair(path)} ${dir}`
      )}`
    : "";
  const offsetFrag = plan.offset ? `OFFSET ${plan.offset}` : "";

  const joinFrags = joinsToString(plan.joins);
  const whereFrag = plan.filter ? `WHERE ${exprToString(plan.filter)}` : "";

  const groupByItems = plan.groupBy.map(_.unary(namepathToQuotedPair));
  const groupByFrag = groupByItems.length > 0 ? `GROUP BY ${groupByItems}` : "";

  return `
  SELECT ${selectFrag}
  FROM ${plan.entry.toLowerCase()} AS "${plan.entry}"
  ${joinFrags}
  ${whereFrag}
  ${groupByFrag}
  ${orderFrag}
  ${limitFrag}
  ${offsetFrag} 
  `;
}
