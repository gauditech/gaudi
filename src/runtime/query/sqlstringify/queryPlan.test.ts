import { format } from "sql-formatter";

import { QueryPlan } from "./queryPlan";
import { queryPlanToString } from "./stringify";

import { compileToOldSpec } from "@src/compiler";
import { compose } from "@src/composer/composer";
import { CustomManyEndpointDef, FetchOneAction, QueryDef } from "@src/types/definition";

/**
 * Define a query wrapper
 */

function makeTestQuery(models: string, query: string): QueryDef {
  const bp = `
  ${models}
  
  model TestHelperModel {}
  entrypoint TestHelperModel {
    custom endpoint {
      method GET
      cardinality many
      path "/test"
      action {
        fetch as q {
          ${query}
        }
      }
    }
  }
  `;
  const def = compose(compileToOldSpec(bp));
  const endpoint = def.entrypoints[0].endpoints[0] as CustomManyEndpointDef;
  const action = endpoint.actions[0] as FetchOneAction;
  return action.query;
}

describe("Query plan", () => {
  function setup() {
    const modelBp = `
    model Org {
      field name { type string }
      relation repos { from Repo, through org }
  
      computed total_issues { count(repos.issues) }
    }
    
    model Repo {
      reference org { to Org }
      field name { type string }
      field ref_number { type integer }
      relation issues { from Issues, through repo }
  
      computed org_name { org.name }
      computed org_total_issues { org.total_issues + 1 }
    }
    
    model Issues {
      field name { type string }
      reference repo { to Repo }
  
      computed repo_name { repo.name }
    }
  `;

    const queryBp = `
  query {
    from Org.repos as o.r,
    filter {
      count(r.issues) // fixme it should work without '.id' as well
      + count(r.issues.repo.id) + sum(r.issues.id) > count(r.issues.repo_name) + r.ref_number
      + count(o.repos.id)
    },
    select { id, name, org_name }
  }
`;

    return makeTestQuery(modelBp, queryBp);
  }

  it("composes a query", () => {
    const query = setup();
    expect(query).toMatchSnapshot();
  });
  it("stringifies the query plan", () => {
    // const query = setup();
    const qstr = queryPlanToString(QP);
    const final = format(qstr, { language: "postgresql" });

    expect(final).toMatchInlineSnapshot(`
      "SELECT
        "Org.repos"."id" AS "id",
        "Org.repos"."name" AS "name",
        "Org.repos.org"."name" AS "org_name"
      FROM
        org AS "Org"
        JOIN repo AS "Org.repos" ON "Org"."id" = "Org.repos"."org_id"
        JOIN org AS "Org.repos.org" ON "Org.repos"."org_id" = "Org.repos.org"."id"
        JOIN (
          SELECT
            "Repo"."id" AS "__join_connection",
            count("Repo.issues"."id") AS "result"
          FROM
            repo AS "Repo"
            JOIN issue AS "Repo.issues" ON "Repo"."id" = "Repo.issues"."repo_id"
          GROUP BY
            "Repo"."id"
        ) AS "Org.repos.COUNT.issues.id" ON "Org.repos"."id" = "Org.repos.COUNT.issues.id"."__join_connection"
        JOIN (
          SELECT
            "Repo"."id" AS "__join_connection",
            count("Repo.issues.repo"."id") AS "result"
          FROM
            repo AS "Repo"
            JOIN issue AS "Repo.issues" ON "Repo"."id" = "Repo.issues"."repo_id"
            JOIN repo AS "Repo.issues.repo" ON "Repo.issues"."repo_id" = "Repo.issues.repo"."id"
          GROUP BY
            "Repo"."id"
        ) AS "Org.repos.COUNT.issues.repo.id" ON "Org.repos"."id" = "Org.repos.COUNT.issues.repo.id"."__join_connection"
        JOIN (
          SELECT
            "Repo"."id" AS "__join_connection",
            sum("Repo.issues"."id") AS "result"
          FROM
            repo AS "Repo"
            JOIN issue AS "Repo.issues" ON "Repo"."id" = "Repo.issues"."repo_id"
          GROUP BY
            "Repo"."id"
        ) AS "Org.repos.SUM.Repo.issues.id" ON "Org.repos"."id" = "Org.repos.SUM.Repo.issues.id"."__join_connection"
        JOIN (
          SELECT
            "Repo"."id" AS "__join_connection",
            count("Repo.issues.repo"."name") AS "result"
          FROM
            repo AS "Repo"
            JOIN issue AS "Repo.issues" ON "Repo"."id" = "Repo.issues"."repo_id"
            JOIN repo AS "Repo.issues.repo" ON "Repo.issues"."repo_id" = "Repo.issues.repo"."id"
          GROUP BY
            "Repo"."id"
        ) AS "Org.repos.COUNT.issues.repo.name" ON "Org.repos"."id" = "Org.repos.COUNT.issues.repo.name"."__join_connection"
        JOIN (
          SELECT
            "Org"."id" AS "__join_connection",
            count("Org.repos"."id") AS "result"
          FROM
            org AS "Org"
            JOIN repo AS "Org.repos" ON "Org"."id" = "Org.repos"."org_id"
          GROUP BY
            "Org"."id"
        ) AS "Org.COUNT.repos.id" ON "Org"."id" = "Org.COUNT.repos.id"."__join_connection"
      WHERE
        (
          "Org.repos.COUNT.issues.id"."result" + (
            "Org.repos.COUNT.issues.repo.id"."result" + "Org.repos.SUM.issues.id"."result"
          )
        ) > (
          "Org.repos.COUNT.issues.repo_name"."result" + (
            "Org.repos"."ref_number" + "Org.COUNT.repos.id"."result"
          )
        )"
    `);
  });
});

const QP: QueryPlan = {
  /**
   * Start main query, Org->repos->org
   */
  entry: "Org",
  groupBy: [],
  filter: {
    kind: "function",
    fnName: ">",
    args: [
      // count(r.issues) + count(r.issues.repo.id) + sum(r.issues.id)
      {
        kind: "function",
        fnName: "+",
        args: [
          // count(r.issues)
          {
            kind: "alias",
            value: ["Org", "repos", "COUNT", "issues", "id", "result"],
          },
          // count(r.issues.repo.id) + sum(r.issues.id)
          {
            kind: "function",
            fnName: "+",
            args: [
              // count(r.issues.repo.id)
              {
                kind: "alias",
                value: ["Org", "repos", "COUNT", "issues", "repo", "id", "result"],
              },
              // sum(r.issues.id)
              {
                kind: "alias",
                value: ["Org", "repos", "SUM", "issues", "id", "result"],
              },
            ],
          },
        ],
      },
      // count(r.issues.repo_name) + r.ref_number + count(o.repos.id)
      {
        kind: "function",
        fnName: "+",
        args: [
          // count(r.issues.repo_name)
          {
            kind: "alias",
            value: ["Org", "repos", "COUNT", "issues", "repo_name", "result"],
          },
          // r.ref_number + count(o.repos.id)
          {
            kind: "function",
            fnName: "+",
            args: [
              // r.ref_number
              {
                kind: "alias",
                value: ["Org", "repos", "ref_number"],
              },
              // count(o.repos.id)
              {
                kind: "alias",
                value: ["Org", "COUNT", "repos", "id", "result"],
              },
            ],
          },
        ],
      },
    ],
  },
  select: {
    id: { kind: "alias", value: ["Org", "repos", "id"] },
    name: { kind: "alias", value: ["Org", "repos", "name"] },
    org_name: { kind: "alias", value: ["Org", "repos", "org", "name"] },
  },
  joins: [
    /**
     * Main query, join Org->repos
     */
    {
      kind: "inline",
      joinType: "inner",
      joinOn: [
        ["Org", "id"],
        ["Org", "repos", "org_id"],
      ],
      modelName: "Repo",
      target: "repos",
      namePath: ["Org", "repos"],
    },
    /**
     * Main query, join [Org->]repos->org
     */
    {
      kind: "inline",
      joinType: "inner",
      joinOn: [
        ["Org", "repos", "org_id"],
        ["Org", "repos", "org", "id"],
      ],
      modelName: "Org",
      namePath: ["Org", "repos", "org"],
      target: "org",
    },
    /**
     * Aggregate query Org->repos->count(issues.id)
     */
    {
      kind: "subquery",
      joinType: "inner",
      namePath: ["Org", "repos", "COUNT", "issues", "id"],
      joinOn: [
        ["Org", "repos", "id"],
        ["Org", "repos", "COUNT", "issues", "id", "__join_connection"],
      ],
      plan: {
        entry: "Repo",
        groupBy: [["Repo", "id"]],
        select: {
          __join_connection: { kind: "alias", value: ["Repo", "id"] },
          result: {
            kind: "function",
            fnName: "count",
            args: [{ kind: "alias", value: ["Repo", "issues", "id"] }],
          },
        },
        /**
         * Inside aggregate, Join Repo->issues
         */
        joins: [
          {
            kind: "inline",
            joinType: "left",
            joinOn: [
              ["Repo", "id"],
              ["Repo", "issues", "repo_id"],
            ],
            modelName: "Issue",
            target: "issues",
            namePath: ["Repo", "issues"],
          },
        ],
      },
    },
    /**
     * Aggregate query Org->repos->count(issues.repo.id)
     */
    {
      kind: "subquery",
      joinType: "inner",
      namePath: ["Org", "repos", "COUNT", "issues", "repo", "id"],
      joinOn: [
        ["Org", "repos", "id"],
        ["Org", "repos", "COUNT", "issues", "repo", "id", "__join_connection"],
      ],
      plan: {
        entry: "Repo",
        groupBy: [["Repo", "id"]],
        select: {
          __join_connection: { kind: "alias", value: ["Repo", "id"] },
          result: {
            kind: "function",
            fnName: "count",
            args: [{ kind: "alias", value: ["Repo", "issues", "repo", "id"] }],
          },
        },
        /**
         * Inside aggregate, Join Repo->issues
         */
        joins: [
          {
            kind: "inline",
            joinType: "left",
            joinOn: [
              ["Repo", "id"],
              ["Repo", "issues", "repo_id"],
            ],
            modelName: "Issue",
            target: "issues",
            namePath: ["Repo", "issues"],
          },
          /**
           * Inside aggregate, Join [Repo->]issues->repo
           */
          {
            kind: "inline",
            joinType: "left",
            joinOn: [
              ["Repo", "issues", "repo_id"],
              ["Repo", "issues", "repo", "id"],
            ],
            modelName: "Repo",
            target: "repo",
            namePath: ["Repo", "issues", "repo"],
          },
        ],
      },
    },
    /**
     * Aggregate query Org->repos->sum(issues.id)
     */
    {
      kind: "subquery",
      joinType: "inner",
      namePath: ["Org", "repos", "SUM", "Repo", "issues", "id"],
      joinOn: [
        ["Org", "repos", "id"],
        ["Org", "repos", "SUM", "Repo", "issues", "id", "__join_connection"],
      ],
      plan: {
        entry: "Repo",
        groupBy: [["Repo", "id"]],
        select: {
          __join_connection: { kind: "alias", value: ["Repo", "id"] },
          result: {
            kind: "function",
            fnName: "sum",
            args: [{ kind: "alias", value: ["Repo", "issues", "id"] }],
          },
        },
        /**
         * Inside aggregate, Join Repo->issues
         */
        joins: [
          {
            kind: "inline",
            joinType: "left",
            joinOn: [
              ["Repo", "id"],
              ["Repo", "issues", "repo_id"],
            ],
            modelName: "Issue",
            target: "issues",
            namePath: ["Repo", "issues"],
          },
        ],
      },
    },
    /**
     * Aggregate query Org->repos->count(issues.repo_name)
     */
    {
      kind: "subquery",
      joinType: "inner",
      namePath: ["Org", "repos", "COUNT", "issues", "repo", "name"],
      joinOn: [
        ["Org", "repos", "id"],
        ["Org", "repos", "COUNT", "issues", "repo", "name", "__join_connection"],
      ],
      plan: {
        entry: "Repo",
        groupBy: [["Repo", "id"]],
        select: {
          __join_connection: { kind: "alias", value: ["Repo", "id"] },
          result: {
            kind: "function",
            fnName: "count",
            args: [{ kind: "alias", value: ["Repo", "issues", "repo", "name"] }],
          },
        },
        /**
         * Inside aggregate, Join Repo->issues
         */
        joins: [
          {
            kind: "inline",
            joinType: "left",
            joinOn: [
              ["Repo", "id"],
              ["Repo", "issues", "repo_id"],
            ],
            modelName: "Issue",
            target: "issues",
            namePath: ["Repo", "issues"],
          },
          /**
           * Inside aggregate, Join [Repo->]issues->repo
           */
          {
            kind: "inline",
            joinType: "left",
            joinOn: [
              ["Repo", "issues", "repo_id"],
              ["Repo", "issues", "repo", "id"],
            ],
            modelName: "Repo",
            target: "repo",
            namePath: ["Repo", "issues", "repo"],
          },
        ],
      },
    },
    /**
     * Join Org->count(Org.repos.id)
     */
    {
      kind: "subquery",
      joinType: "left",
      namePath: ["Org", "COUNT", "repos", "id"],
      joinOn: [
        ["Org", "id"],
        ["Org", "COUNT", "repos", "id", "__join_connection"],
      ],
      plan: {
        /**
         * Aggregate query Org->repos
         */
        entry: "Org",
        groupBy: [["Org", "id"]],
        select: {
          __join_connection: { kind: "alias", value: ["Org", "id"] },
          result: {
            kind: "function",
            fnName: "count",
            args: [{ kind: "alias", value: ["Org", "repos", "id"] }],
          },
        },
        joins: [
          {
            kind: "inline",
            joinType: "left",
            joinOn: [
              ["Org", "id"],
              ["Org", "repos", "org_id"],
            ],
            modelName: "Repo",
            namePath: ["Org", "repos"],
            target: "repos",
          },
        ],
      },
    },
  ],
};
