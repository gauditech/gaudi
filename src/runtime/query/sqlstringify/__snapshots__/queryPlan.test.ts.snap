// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query plan composes a query 1`] = `
{
  "filter": {
    "args": [
      {
        "args": [
          {
            "args": [
              {
                "args": [
                  {
                    "kind": "alias",
                    "namePath": [
                      "Org",
                      "repos",
                      "issues",
                    ],
                  },
                ],
                "kind": "function",
                "name": "count",
              },
              {
                "args": [
                  {
                    "kind": "alias",
                    "namePath": [
                      "Org",
                      "repos",
                      "issues",
                      "repo",
                      "id",
                    ],
                  },
                ],
                "kind": "function",
                "name": "count",
              },
            ],
            "kind": "function",
            "name": "+",
          },
          {
            "args": [
              {
                "kind": "alias",
                "namePath": [
                  "Org",
                  "repos",
                  "issues",
                  "id",
                ],
              },
            ],
            "kind": "function",
            "name": "sum",
          },
        ],
        "kind": "function",
        "name": "+",
      },
      {
        "args": [
          {
            "args": [
              {
                "args": [
                  {
                    "kind": "alias",
                    "namePath": [
                      "Org",
                      "repos",
                      "issues",
                      "repo_name",
                    ],
                  },
                ],
                "kind": "function",
                "name": "count",
              },
              {
                "kind": "alias",
                "namePath": [
                  "Org",
                  "repos",
                  "ref_number",
                ],
              },
            ],
            "kind": "function",
            "name": "+",
          },
          {
            "args": [
              {
                "kind": "alias",
                "namePath": [
                  "Org",
                  "repos",
                  "id",
                ],
              },
            ],
            "kind": "function",
            "name": "count",
          },
        ],
        "kind": "function",
        "name": "+",
      },
    ],
    "kind": "function",
    "name": ">",
  },
  "fromPath": [
    "Org",
    "repos",
  ],
  "kind": "query",
  "limit": undefined,
  "modelRefKey": "Org",
  "name": "$query",
  "offset": undefined,
  "orderBy": undefined,
  "refKey": "N/A",
  "retType": "Repo",
  "select": [
    {
      "alias": "id",
      "kind": "field",
      "name": "id",
      "namePath": [
        "Org",
        "repos",
        "id",
      ],
      "refKey": "Repo.id",
    },
    {
      "alias": "name",
      "kind": "field",
      "name": "name",
      "namePath": [
        "Org",
        "repos",
        "name",
      ],
      "refKey": "Repo.name",
    },
    {
      "alias": "org_name",
      "kind": "computed",
      "name": "org_name",
      "namePath": [
        "Org",
        "repos",
        "org_name",
      ],
      "refKey": "Repo.org_name",
    },
  ],
}
`;
